File: src/main/java/com/api/sistema_penal/service/SentencaService.java
package com.api.sistema_penal.service;

import com.api.sistema_penal.api.dto.sentenca.*;
import com.api.sistema_penal.domain.entity.Processo;
import com.api.sistema_penal.domain.entity.Sentenca;
import com.api.sistema_penal.domain.entity.Sentenca.TipoDecisao;
import com.api.sistema_penal.domain.repository.ProcessoRepository;
import com.api.sistema_penal.domain.repository.SentencaRepository;
import com.api.sistema_penal.exception.BusinessException;
import com.api.sistema_penal.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class SentencaService {

    private final SentencaRepository sentencaRepository;
    private final ProcessoRepository processoRepository;

    @Transactional(readOnly = true)
    public Page<SentencaSummaryResponse> listar(Pageable pageable) {
        return sentencaRepository.findAll(pageable).map(SentencaSummaryResponse::from);
    }

    @Transactional(readOnly = true)
    public Page<SentencaSummaryResponse> listarPorTipoDecisao(TipoDecisao tipo, Pageable pageable) {
        return sentencaRepository.findByTipoDecisao(tipo, pageable).map(SentencaSummaryResponse::from);
    }

    @Transactional(readOnly = true)
    public Page<SentencaSummaryResponse> listarPorPeriodo(LocalDate inicio, LocalDate fim, Pageable pageable) {
        return sentencaRepository.findByPeriodo(inicio, fim, pageable).map(SentencaSummaryResponse::from);
    }

    @Transactional(readOnly = true)
    public Page<SentencaSummaryResponse> listarPorTipoCrime(UUID tipoCrimeId, Pageable pageable) {
        // Funcionalidade não disponível - Processo não possui campo tipoCrime
        return Page.empty(pageable);
    }

    @Transactional(readOnly = true)
    public Page<SentencaSummaryResponse> buscarJurisprudencia(String termo, Pageable pageable) {
        return sentencaRepository.buscarPorTexto(termo, pageable).map(SentencaSummaryResponse::from);
    }

    @Transactional(readOnly = true)
    public SentencaResponse buscarPorId(UUID id) {
        Sentenca sentenca = sentencaRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Sentença", id));
        return SentencaResponse.from(sentenca);
    }

    @Transactional(readOnly = true)
    public SentencaResponse buscarPorProcesso(UUID processoId) {
        Sentenca sentenca = sentencaRepository.findByProcessoId(processoId)
                .orElseThrow(() -> new ResourceNotFoundException("Sentença do processo " + processoId));
        return SentencaResponse.from(sentenca);
    }

    @Transactional
    public SentencaResponse criar(SentencaRequest request) {
        if (sentencaRepository.findByProcessoId(request.processoId()).isPresent()) {
            throw new BusinessException("Processo já possui sentença");
        }

        Processo processo = processoRepository.findById(request.processoId())
                .orElseThrow(() -> new ResourceNotFoundException("Processo", request.processoId()));

        Sentenca sentenca = Sentenca.builder()
                .processo(processo)
                .tipoDecisao(request.tipoDecisao())
                .penaMeses(request.penaMeses())
                .tipoPena(request.tipoPena())
                .regime(request.regime())
                .dataSentenca(request.dataSentenca())
                .ementa(request.ementa())
                .fundamentacao(request.fundamentacao())
                .dispositivo(request.dispositivo())
                .juizNome(request.juizNome())
                .circunstancias(request.circunstancias() != null ? request.circunstancias() : new HashMap<>())
                .build();

        processo.setStatus(Processo.StatusProcesso.SENTENCIADO);
        processoRepository.save(processo);

        return SentencaResponse.from(sentencaRepository.save(sentenca));
    }

    @Transactional
    public SentencaResponse atualizar(UUID id, SentencaRequest request) {
        Sentenca sentenca = sentencaRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Sentença", id));

        sentenca.setTipoDecisao(request.tipoDecisao());
        sentenca.setPenaMeses(request.penaMeses());
        sentenca.setTipoPena(request.tipoPena());
        sentenca.setRegime(request.regime());
        sentenca.setDataSentenca(request.dataSentenca());
        sentenca.setEmenta(request.ementa());
        sentenca.setFundamentacao(request.fundamentacao());
        sentenca.setDispositivo(request.dispositivo());
        sentenca.setJuizNome(request.juizNome());
        if (request.circunstancias() != null) {
            sentenca.setCircunstancias(request.circunstancias());
        }

        return SentencaResponse.from(sentencaRepository.save(sentenca));
    }

    @Transactional
    public void marcarTransitadoJulgado(UUID id) {
        Sentenca sentenca = sentencaRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Sentença", id));
        
        sentenca.setTransitadoJulgado(true);
        sentencaRepository.save(sentenca);

        Processo processo = sentenca.getProcesso();
        processo.setStatus(Processo.StatusProcesso.TRANSITADO_JULGADO);
        processo.setDataEncerramento(LocalDate.now());
        processoRepository.save(processo);
    }

    @Transactional(readOnly = true)
    public Map<String, Object> estatisticas() {
        Map<String, Object> stats = new HashMap<>();

        stats.put("porTipoDecisao", sentencaRepository.countByTipoDecisao().stream()
                .collect(Collectors.toMap(
                        arr -> ((TipoDecisao) arr[0]).name(),
                        arr -> arr[1]
                )));

        stats.put("total", sentencaRepository.count());

        return stats;
    }

    @Transactional(readOnly = true)
    public Double mediaPenaPorTipoCrime(UUID tipoCrimeId) {
        // Funcionalidade não disponível - Processo não possui campo tipoCrime
        return null;
    }
}

File: frontend/src/services/sentenca.service.ts
import api from '@/lib/api';
import { Page } from '@/types';

export interface Sentenca {
  id: string;
  processoNumero: string;
  tipoDecisao: string;
  tipoCrimeNome?: string;
  penaMeses?: number;
  tipoPena?: string;
  regime?: string;
  dataSentenca?: string;
  ementa?: string;
  fundamentacao?: string;
  dispositivo?: string;
  narrativa?: string;
  juizNome?: string;
  transitadoJulgado?: boolean;
  createdAt?: string;
}

export interface SentencaEstatisticas {
  total?: number;
  condenacoes?: number;
  absolicoes?: number;
  mediaPena?: number;
}

const sentencaService = {
  listar(params?: { tipoDecisao?: string; tipoCrimeId?: string }, page = 0, size = 10) {
    const queryParams = new URLSearchParams();
    queryParams.append('page', page.toString());
    queryParams.append('size', size.toString());
    
    if (params?.tipoDecisao) {
      queryParams.append('tipoDecisao', params.tipoDecisao);
    }
    if (params?.tipoCrimeId) {
      queryParams.append('tipoCrimeId', params.tipoCrimeId);
    }
    
    return api.get<Page<Sentenca>>(`/sentencas?${queryParams.toString()}`).then(res => res.data);
  },

  buscarJurisprudencia(termoBusca: string, page = 0, size = 10) {
    const queryParams = new URLSearchParams();
    queryParams.append('q', termoBusca);
    queryParams.append('page', page.toString());
    queryParams.append('size', size.toString());
    
    return api.get<Page<Sentenca>>(`/sentencas/jurisprudencia?${queryParams.toString()}`).then(res => res.data);
  },

  buscarPorId(id: string) {
    return api.get<Sentenca>(`/sentencas/${id}`).then(res => res.data);
  },

  buscarPorProcesso(processoId: string) {
    return api.get<Sentenca>(`/sentencas/processo/${processoId}`).then(res => res.data);
  },

  estatisticas() {
    return api.get<SentencaEstatisticas>('/sentencas/estatisticas').then(res => res.data);
  },

  criar(data: Partial<Sentenca>) {
    return api.post<Sentenca>('/sentencas', data).then(res => res.data);
  },

  atualizar(id: string, data: Partial<Sentenca>) {
    return api.put<Sentenca>(`/sentencas/${id}`, data).then(res => res.data);
  },

  transitarJulgado(id: string) {
    return api.patch<{ message: string }>(`/sentencas/${id}/transito-julgado`).then(res => res.data);
  },
};

export { sentencaService };
